File: frontend/src/api/pings.ts
import { api } from './client'

/** In-app work routing. Read state is per person; DONE is shared across the
 *  roster (server-derived), so `done_by` names whoever closed it first. */

export interface PingRecipient {
  user_id: string
  name: string
  email: string
  read_at: string | null
  completed_at: string | null
}

/** Identifying summary of the request a ping is about (spec section 4.1). It
 *  renders even when the viewer cannot open the request; `visible` is what
 *  gates the deep link, or a link that 403s reads as broken software. */
export interface RequestRef {
  id: string
  number: string
  status: string
  division_name: string | null
  requestor_name: string | null
  total_cost: string | null
  visible: boolean
}

export interface PingSummary {
  id: string
  note: string
  sender: { id: string; name: string }
  created_at: string
  request_id: string | null
  request: RequestRef | null
  recipients: PingRecipient[]
  reply_count: number
  unread_replies: number
  unread_for_me: boolean
  last_activity_at: string
  completed_at: string | null
  done_by: string | null
  /** Every reply's note, concatenated -- what search reaches beyond the root
   *  note and the sender/recipient names. */
  search_blob: string
}

export interface PingDetail extends PingSummary {
  replies: {
    id: string
    note: string
    created_at: string
    sender: { id: string; name: string }
  }[]
}

export interface DirectoryUser {
  id: string
  name: string
  email: string
  roles: string[]
  division_name: string | null
}

export interface PingCreate {
  recipient_ids: string[]
  note: string
  request_id?: string | null
}

export async function pingUnreadCount(): Promise<number> {
  const res = await api<{ count: number }>('/pings/unread_count')
  return res.count
}

export async function listPings(box: 'inbox' | 'sent'): Promise<PingSummary[]> {
  const res = await api<{ pings: PingSummary[] }>(`/pings?box=${box}`)
  return res.pings
}

export async function getPing(id: string): Promise<PingDetail> {
  const res = await api<{ ping: PingDetail }>(`/pings/${id}`)
  return res.ping
}

export async function createPing(body: PingCreate): Promise<PingDetail> {
  const res = await api<{ ping: PingDetail }>('/pings', { method: 'POST', body })
  return res.ping
}

export async function replyToPing(id: string, note: string): Promise<PingDetail> {
  const res = await api<{ ping: PingDetail }>(`/pings/${id}/reply`, {
    method: 'POST',
    body: { note },
  })
  return res.ping
}

export async function markPingDone(id: string): Promise<PingDetail> {
  const res = await api<{ ping: PingDetail }>(`/pings/${id}/done`, { method: 'POST' })
  return res.ping
}

export async function reopenPing(id: string): Promise<PingDetail> {
  const res = await api<{ ping: PingDetail }>(`/pings/${id}/reopen`, { method: 'POST' })
  return res.ping
}

export async function pingDirectory(): Promise<DirectoryUser[]> {
  const res = await api<{ users: DirectoryUser[] }>('/pings/directory')
  return res.users
}

/** The requestor plus, when pending, the current level's eligible approvers
 *  (FINANCE once approved) -- ranked first in the New Ping modal. */
export async function pingSuggestions(requestId: string): Promise<DirectoryUser[]> {
  const res = await api<{ users: DirectoryUser[] }>(
    `/pings/suggestions?request_id=${encodeURIComponent(requestId)}`,
  )
  return res.users
}

File: backend/app/blueprints/pings.py
"""Pings. Thin routes; everything lives in ping_service.

No require_roles: every role pings, and the directory is unscoped by design
(spec section 5). login_required alone is the gate. Static paths (unread_count,
directory, suggestions) are declared before /<ping_id> so Flask never treats
them as ids.
"""
from flask import Blueprint, jsonify, request
from flask_login import current_user, login_required

from app.schemas.pings import PingCreateIn, PingReplyIn
from app.services import ping_service
from app.services.errors import ServiceError

bp = Blueprint("pings", __name__, url_prefix="/api/pings")


@bp.get("/unread_count")
@login_required
def unread_count():
    return jsonify(count=ping_service.unread_count(current_user))


@bp.get("")
@login_required
def list_pings():
    box = request.args.get("box") or "inbox"
    return jsonify(pings=ping_service.list_pings(current_user, box))


@bp.get("/directory")
@login_required
def directory():
    return jsonify(users=ping_service.directory())


@bp.get("/suggestions")
@login_required
def suggestions():
    request_id = request.args.get("request_id")
    if not request_id:
        raise ServiceError("request_id is required.", 400)
    return jsonify(users=ping_service.suggested_recipients(current_user, request_id))


@bp.get("/<ping_id>")
@login_required
def get_ping(ping_id):
    return jsonify(ping=ping_service.get_ping(current_user, ping_id))


@bp.post("")
@login_required
def create_ping():
    data = PingCreateIn(**(request.get_json(silent=True) or {}))
    ping = ping_service.create_ping(
        current_user, recipient_ids=data.recipient_ids, note=data.note,
        request_id=data.request_id)
    return jsonify(ping=ping_service.get_ping(current_user, ping.id))


@bp.post("/<ping_id>/reply")
@login_required
def reply(ping_id):
    data = PingReplyIn(**(request.get_json(silent=True) or {}))
    ping_service.reply(current_user, ping_id, data.note)
    return jsonify(ping=ping_service.get_ping(current_user, ping_id))


@bp.post("/<ping_id>/done")
@login_required
def done(ping_id):
    return jsonify(ping=ping_service.complete_ping(current_user, ping_id))


@bp.post("/<ping_id>/reopen")
@login_required
def reopen(ping_id):
    return jsonify(ping=ping_service.reopen_ping(current_user, ping_id))
